# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/quota/QuotaManager.h
# Commit: 8380179aa33f
# Full Hash: 8380179aa33f282702b219de98afa653539edac1
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2024-10-10 09:22:33
# Regressor Bug: 1866402
# File Overlap Count: 1
# Description:
#   Bug 1866402 - Implement origin initialization status tracking on the owning thread; r=dom-storage-reviewers,jari
#   
#   Origin initialization tracking on the owning thread is needed for covering
#   origin initialization in QuotaManager::OpenClientDirectory.
#   
# ==============================================================================

diff -r e18f8613cd79 -r 8380179aa33f dom/quota/QuotaManager.h
--- a/dom/quota/QuotaManager.h	Wed Oct 09 22:16:51 2024 +0000
+++ b/dom/quota/QuotaManager.h	Wed Oct 09 22:16:52 2024 +0000
@@ -25,6 +25,7 @@
 #include "mozilla/dom/quota/CommonMetadata.h"
 #include "mozilla/dom/quota/DirectoryLockCategory.h"
 #include "mozilla/dom/quota/ForwardDecls.h"
+#include "mozilla/dom/quota/HashKeys.h"
 #include "mozilla/dom/quota/InitializationTypes.h"
 #include "mozilla/dom/quota/NotifyUtils.h"
 #include "mozilla/dom/quota/OriginOperationCallbacks.h"
@@ -84,6 +85,7 @@
   friend class CanonicalQuotaObject;
   friend class ClearStorageOp;
   friend class DirectoryLockImpl;
+  friend class FinalizeOriginEvictionOp;
   friend class GroupInfo;
   friend class InitOp;
   friend class InitTemporaryStorageOp;
@@ -375,6 +377,8 @@
   RefPtr<BoolPromise> PersistentOriginInitialized(
       const PrincipalInfo& aPrincipalInfo);
 
+  bool IsPersistentOriginInitialized(const PrincipalInfo& aPrincipalInfo);
+
   bool IsPersistentOriginInitializedInternal(
       const OriginMetadata& aOriginMetadata) const;
 
@@ -394,6 +398,9 @@
   RefPtr<BoolPromise> TemporaryOriginInitialized(
       PersistenceType aPersistenceType, const PrincipalInfo& aPrincipalInfo);
 
+  bool IsTemporaryOriginInitialized(PersistenceType aPersistenceType,
+                                    const PrincipalInfo& aPrincipalInfo);
+
   bool IsTemporaryOriginInitializedInternal(
       const OriginMetadata& aOriginMetadata) const;
 
@@ -743,6 +750,17 @@
 
   void ClearDirectoryLockTables();
 
+  void NoteInitializedOrigin(PersistenceType aPersistenceType,
+                             const nsACString& aOrigin);
+
+  void NoteUninitializedOrigins(
+      const OriginMetadataArray& aOriginMetadataArray);
+
+  void NoteUninitializedRepository(PersistenceType aPersistenceType);
+
+  bool IsOriginInitialized(PersistenceType aPersistenceType,
+                           const nsACString& aOrigin) const;
+
   bool IsSanitizedOriginValid(const nsACString& aSanitizedOrigin);
 
   Result<nsCString, nsresult> EnsureStorageOriginFromOrigin(
@@ -821,6 +839,10 @@
   DirectoryLockTable mDefaultDirectoryLockTable;
   DirectoryLockTable mPrivateDirectoryLockTable;
 
+  using BoolArray = AutoTArray<bool, PERSISTENCE_TYPE_INVALID>;
+  nsTHashMap<nsCStringHashKeyWithDisabledMemmove, BoolArray>
+      mInitializedOrigins;
+
   // A list of all successfully initialized persistent origins. This list isn't
   // protected by any mutex but it is only ever touched on the IO thread.
   nsTArray<nsCString> mInitializedOriginsInternal;